native/membrane_videocompositor_opengl_rust/src/framebuffers.rs

```diff
@@ -0,0 +1,171 @@
+//! Structures for managing various OpenGL render targets
+use glad_gles2::gl;
+
+/// An abstraction for an individual [framebuffer object](<https://www.khronos.org/opengl/wiki/Framebuffer_Object>) with an attached [renderbuffer](<https://www.khronos.org/opengl/wiki/Renderbuffer_Object>).
+struct FramebufferObject {
+    width: usize,
+    height: usize,
+    id: gl::GLuint,
+    renderbuffer_id: gl::GLuint,
+    _internal_format: gl::GLenum, // FIXME these should be custom enums instead of GLenum random ints
+    output_format: gl::GLenum,
+    output_type: gl::GLenum,
+}
+
+impl FramebufferObject {
+    /// Create a new instance.
+    ///  * `width` and `height` should be given in pixels.
+    ///  * `internal_format` is the format used by the renderbuffer internally (e.g `gl::RGB8` or `gl::R8`)
+    ///  * `output_format` and `output_type` represent the type and format we expect to get out when reading from the framebuffer (e.g. `gl::RGB` for `format` and `gl::UNSIGNED_BYTE` for `output_type`)
+    fn new(
+        width: usize,
+        height: usize,
+        internal_format: gl::GLenum,
+        output_format: gl::GLenum,
+        output_type: gl::GLenum,
+    ) -> Self {
+        let mut id = 0;
+        let mut renderbuffer_id = 0;
+
+        unsafe {
+            gl::GenFramebuffers(1, &mut id);
+            gl::BindFramebuffer(gl::FRAMEBUFFER, id);
+
+            gl::GenRenderbuffers(1, &mut renderbuffer_id);
+
+            gl::BindRenderbuffer(gl::RENDERBUFFER, renderbuffer_id);
+            gl::RenderbufferStorage(
+                gl::RENDERBUFFER,
+                internal_format,
+                width as i32,
+                height as i32,
+            );
+            gl::FramebufferRenderbuffer(
+                gl::FRAMEBUFFER,
+                gl::COLOR_ATTACHMENT0,
+                gl::RENDERBUFFER,
+                renderbuffer_id,
+            );
+        }
+
+        Self {
+            width,
+            height,
+            id,
+            renderbuffer_id,
+            _internal_format: internal_format,
+            output_format,
+            output_type,
+        }
+    }
+
+    fn bind_for_drawing(&self) {
+        unsafe {
+            gl::BindFramebuffer(gl::DRAW_FRAMEBUFFER, self.id);
+            gl::DrawBuffers(1, [gl::COLOR_ATTACHMENT0].as_ptr());
+            gl::Viewport(0, 0, self.width as i32, self.height as i32);
+        }
+    }
+
+    fn bind_for_reading(&self) {
+        unsafe {
+            gl::BindFramebuffer(gl::READ_FRAMEBUFFER, self.id);
+            gl::ReadBuffer(gl::COLOR_ATTACHMENT0);
+        }
+    }
+
```

Comment:
this API seems very unsafe. I don't know opengl, but it looks like you have to bind before drawing into. it would be more idiomatic then to return some kind of a proxy structure here that would expose the draw/read methods and keep track of the binding status.

Reply:
I think there is no unsafety here, since OpenGL should just throw an error when trying to draw without binding, but the idea of a proxy makes a lot of sense here.

Reply:
how is the error captured though? 😄 won't this just panic the whole machine? even if this is a programmer's error, it would be nice if abstractions prevented writing bad code

Reply:
it won't. you have to call gl::GetError() to check whether errors have occurred. If you don't react to the error the state machine will be in a state you don't expect and may produce weird results, but it won't crash/panic